sql-plugin/src/main/scala/org/apache/spark/sql/rapids/complexTypeCreator.scala

```diff
@@ -0,0 +1,103 @@
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.rapids
+
+import ai.rapids.cudf.ColumnVector
+import com.nvidia.spark.rapids.{GpuColumnVector, GpuExpression, GpuScalar}
+import com.nvidia.spark.rapids.RapidsPluginImplicits.{AutoCloseableArray, ReallyAGpuExpression}
+
+import org.apache.spark.sql.catalyst.analysis.FunctionRegistry.FUNC_ALIAS
+import org.apache.spark.sql.catalyst.analysis.TypeCheckResult
+import org.apache.spark.sql.catalyst.expressions.{EmptyRow, Expression, NamedExpression}
+import org.apache.spark.sql.types.{Metadata, StringType, StructField, StructType}
+import org.apache.spark.sql.vectorized.ColumnarBatch
+
+case class GpuCreateNamedStruct(children: Seq[Expression]) extends GpuExpression {
+  lazy val (nameExprs, valExprs) = children.grouped(2).map {
+    case Seq(name, value) => (name, value)
+  }.toList.unzip
+
+  private lazy val names = nameExprs.map {
+    case g: GpuExpression => g.columnarEval(null)
+    case e => e.eval(EmptyRow)
+  }
+
+  override def nullable: Boolean = false
+
+  override def foldable: Boolean = valExprs.forall(_.foldable)
+
+  override lazy val dataType: StructType = {
+    val fields = names.zip(valExprs).map {
+      case (name, expr) =>
+        val metadata = expr match {
+          case ne: NamedExpression => ne.metadata
+          case _ => Metadata.empty
+        }
+        StructField(name.toString, expr.dataType, expr.nullable, metadata)
+    }
+    StructType(fields)
+  }
+
+  override def checkInputDataTypes(): TypeCheckResult = {
+    if (children.size % 2 != 0) {
+      TypeCheckResult.TypeCheckFailure(s"$prettyName expects an even number of arguments.")
+    } else {
+      val invalidNames = nameExprs.filterNot(e => e.foldable && e.dataType == StringType)
+      if (invalidNames.nonEmpty) {
+        TypeCheckResult.TypeCheckFailure(
+          s"Only foldable ${StringType.catalogString} expressions are allowed to appear at odd" +
+              s" position, got: ${invalidNames.mkString(",")}")
+      } else if (!names.contains(null)) {
+        TypeCheckResult.TypeCheckSuccess
+      } else {
+        TypeCheckResult.TypeCheckFailure("Field name should not be null")
+      }
+    }
+  }
+
+  // There is an alias set at `CreateStruct.create`. If there is an alias,
+  // this is the struct function explicitly called by a user and we should
+  // respect it in the SQL string as `struct(...)`.
+  override def prettyName: String = getTagValue(FUNC_ALIAS).getOrElse("named_struct")
+
+  override def sql: String = getTagValue(FUNC_ALIAS).map { alias =>
+    val childrenSQL = children.indices.filter(_ % 2 == 1).map(children(_).sql).mkString(", ")
+    s"$alias($childrenSQL)"
+  }.getOrElse(super.sql)
+
+  override def columnarEval(batch: ColumnarBatch): Any = {
+    // The names are only used for the type. Here we really just care about the data
+    val columns: Array[ColumnVector] = new Array(valExprs.size)
+    try {
```

Comment:
Suggest using `withResource` here.

Reply:
That does not work for Arrays because it changes the type to be a Seq and I cannot modify it after that.

Reply:
It should work, as there was a specialization added explicitly for the `Array` case.  The pattern with an Array is being used in [GpuParquetScan](https://github.com/NVIDIA/spark-rapids/blob/branch-0.4/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala#L1034) and [MetaUtils](https://github.com/NVIDIA/spark-rapids/blob/branch-0.4/sql-plugin/src/main/scala/com/nvidia/spark/rapids/MetaUtils.scala#L228), for example.

Reply:
I agree, and I think I tried it once, but ran into some compile issues because now the array could be covered by both functions, and my scala skills were just not strong enough to figure out how to make it work properly. I can try again if you want me to.

Reply:
But I could always add in a different name that would probably make it still work.  I'll give it a shot

Reply:
It's not a big deal, the `try` block works so don't waste too much time on it.  I am curious why it didn't work in this instance.  What's being done here doesn't seem that different than other uses of `withResource(new Array.[..` in the code base.

Reply:
Sorry it looks like someone fixed it for me in Arm. Thanks @tgravescs  So I will update it